### ISSUE TITLE
Missing Attribute Types?

### ISSUE THREAD

**Issue by devlinl** [MicrosoftDocs/powerapps-docs#4897]
This page lists the Metadata Objects for Boolean, Picklist, State, and Status, but what about the others? The various numeric types... Date Time, Lookup/Customer/Owner, Text...

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: bf6143f9-6760-18c0-a08f-196c8bce8c2f
* Version Independent ID: 74a79a52-be75-ad56-642c-8a3aa6b5ab0d
* Content: [getEntityMetadata (Client API reference) in model-driven apps - Power Apps](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/xrm-utility/getentitymetadata#attribute-objects)
* Content Source: [powerapps-docs/developer/model-driven-apps/clientapi/reference/Xrm-Utility/getEntityMetadata.md](https://github.com/MicrosoftDocs/powerapps-docs/blob/main/powerapps-docs/developer/model-driven-apps/clientapi/reference/Xrm-Utility/getEntityMetadata.md)
* Service: **powerapps**
* Sub-service: **mda-developer**
* GitHub Login: @adrianorth
* Microsoft Alias: **aorth**

**Comment by bishalgoswami** [comment]
Thanks for your comment, @devlinl. This issue has been assigned to the doc author for review.

**Comment by JimDaly** [comment]
@devlinl 

Thanks for your feedback. Sorry for the delay in responding.

I think that the [Attribute objects](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/xrm-utility/getentitymetadata#attribute-objects) section is pretty clear.

All types return the [Base (AttributeMetadata) columns](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/xrm-utility/getentitymetadata#base-attributemetadata-columns), and the `AttributeType` property tells you what type.

Only the types mentioned include additional properties. For example, some return an `OptionSet`, and this object has some different methods.

This API doesn't seem to return all available [AttributeMetadata Class](https://learn.microsoft.com/en-us/dotnet/api/microsoft.xrm.sdk.metadata.attributemetadata?view=dataverse-sdk-latest) or derived class properties. It appears to be a selected set of properties that have value for client-side scripting. 

Are you seeing that it returns more properties than documented?

**Comment by devlinl** [comment]
You're correct that not all metadata attributes are returned... So, for the Text fields, it doesn't return Max Length, and for the numeric fields, it doesn't return Min Value and Max Value. However, there are definitely several metadata attributes that are returned for many column types which are not documented here.

- Money
    - Precision
    - PrecisionSource (I didn't check what the values for this mean)
- DateTime
    - Behavior (1: User Local, 2: Date Only, 3: Time Zone Independent)
    - Format ("date" or "dateandtime")
- Decimal & Double
    - Precision
- Integer
    - Format (0: None, 1: Duration, 2: Time Zone, 3: Language)
- Lookup
    - Targets

**Comment by JimDaly** [comment]
OK. I see your point.
Pasting this function in the Developer tools console:

```
var entityName = "account";
var attributes = [
  "revenue", 
  "createdon",
  "exchangerate",
  "address1_latitude",
  "opendeals",
  "primarycontactid"];

// Define the success callback function
var successCallback = function (result) {
  // Log the result to the console
  result.Attributes.forEach(attribute => {
    console.log(attribute.LogicalName);
    console.log(` AttributeType=${attribute.AttributeType}`)
    console.log(` LogicalName=${attribute.LogicalName}`)
    // Money, Decimal, & Double have Precision
    if(attribute.Precision){
      console.log(` Precision=${attribute.Precision}`)
    }
    // DateTime has Behavior
    if(attribute.Behavior){
      console.log(` Behavior=${attribute.Behavior}`)
    }
    // DateTime and Integer have Format
    if(attribute.Format){
      console.log(` Format=${attribute.Format}`)
    }
    // Lookups have Targets
    if(attribute.Targets){
      console.log(` Targets=${attribute.Targets}`)
    }

  });
};

// Define the error callback function
var errorCallback = function (error) {
  // Log the error to the console
  console.error(error);
};

// Call the Xrm.Utility.getEntityMetadata function
Xrm.Utility.getEntityMetadata(entityName, attributes).then(successCallback, errorCallback);
```

I get output like this:

```
revenue
   AttributeType=8
   LogicalName=revenue
   Precision=2
createdon
   AttributeType=2
   LogicalName=createdon
   Behavior=1
   Format=dateandtime
exchangerate
   AttributeType=3
   LogicalName=exchangerate
   Precision=10
address1_latitude
   AttributeType=4
   LogicalName=address1_latitude
   Precision=5
opendeals
   AttributeType=5
   LogicalName=opendeals
   Format=0
primarycontactid
   AttributeType=6
   LogicalName=primarycontactid
   Targets=contact
```

This demonstrates that the properties you listed (`Precision`, `Behavior`, `Format` and `Targets`) are returned, and they aren't included in the document.

This doesn't test for others, but it is evidence that the documented properties are incomplete.
I'll take this up with the dev team. IDK why only a few were documented, or perhaps more were added over time.
